### ISSUE TITLE
Always redirecting after login fails, despite logoutData having redirect: false

### ISSUE THREAD

**Issue by darkylmnx** [websanova/vue-auth#613]
In my project I set all `redirect` options to `false`, that includes `loginData.redirect`, `logoutData.redirect`, `registerData.redirect` and root option `authRedirect`.

For my login form to be accessible anywhere, I show it when `$route.query.login == '1'`.

The issue is: When the login fails and creates a 401, your code calls:
- `_parseResponseIntercept`
- which then calls `_processInvalidToken`
- which then calls `_processLogout` with a redirect despite me explicitly saying everywhere `redirect: false` -> https://github.com/websanova/vue-auth/blob/c06925107499d54e0c2b906e454709d370673359/src/auth.js#L229

Why is that? why have a `logoutData.redirect` option if it's not used here?

Now, the deep issue is, it tries to redirect to the same page which would be something like `/anypage?login=1`.
**BUT** vue-router doesn't take query string in account when doing `router.push({ redirect: '/anypage?login=1' })` which leads to be only redirected to `/anypage` which makes my login form always close when the login fails.

I took me more than 4 hours to debug this as nothing in your docs says a "redirect is done after login fails" nor vue-router's docs saying "redirect doesn't take query string in account".

Actually vue-router only takes query string in account if the redirect is a location object and not a string.

**Comment by websanova** [comment]
Not 100% sure I understand.

"When the login fails and creates a 401", not sure by what you mean by "login fails", this redirect should only occur if you try to access a page that requires login and the user is currently logged out.

For the redirect can you not use `{path: '/anypage/', query: :{login: 1}}` format?

**Comment by darkylmnx** [comment]
Well it seems you didn't understand indeed.
I'm not the one redirecting anything, your plugin is.
- When I call `this.$auth.login( ... )` and the server response fails because the user creditals are wrong for example
- Your plugin call a succession of methods listed above and create a redirection on it's own to the same current URL

Practical example:
- my user is on the URL `/home?login=1` and sees a login modal
- on submit, I call `$auth.login( ... )`
- the HTTP call fails with a 401 HTTP code repsonse
- the plugin tries to redirect to `/home?login=1` (I don't even see why you would do that actually)
- it does this with the call `_processLogout({redirect: redirect});` located https://github.com/websanova/vue-auth/blob/c06925107499d54e0c2b906e454709d370673359/src/auth.js#L229

So basically, I'm not doint anything special here, but as I said I set every redirect to fals because I don't need the plugin to handle them. Even if I actually set the redict like you say (which I tried) it doesn't work, check your code but in that particular line you do not check any option to generate the redirect URL -> https://github.com/websanova/vue-auth/blob/c06925107499d54e0c2b906e454709d370673359/src/auth.js#L229

**Comment by darkylmnx** [comment]
@websanova Actually, Posva from vue-router even says the redirect should not be used in the `router.push` method and only used in route records.

> redirect is supposed to be used in route records, it's not meant to be passed to push: https://router.vuejs.org/guide/essentials/redirect-and-alias.html#redirect 

https://github.com/vuejs/vue-router/issues/3374

**Comment by websanova** [comment]
Well, the parameter just happens to be called "redirect", but it is in fact doing a router.push. That redirect param is for after a success login or whatever, so I think it's correct, perhaps the naming is a bit misleading of you're mapping it directly to vue-roiter.

**Comment by websanova** [comment]
As for dealing with the 401, I think you should be able to control that via the http driver.

https://github.com/websanova/vue-auth/blob/master/src/drivers/http/axios.1.x.js
https://github.com/websanova/vue-auth/blob/master/src/drivers/http/vue-resource.1.x.js

There is an `invalidToken` method you can override.

**Comment by darkylmnx** [comment]
This code shows that you are passing an object with a reidrect key to the router push method so I don't see how it can be correct when the author explicitly says:

> redirect is supposed to be used in route records, it's not meant to be passed to push: https://router.vuejs.org/guide/essentials/redirect-and-alias.html#redirect

Here's your code for the routerGo in image:
![image](https://user-images.githubusercontent.com/1426357/99385845-d9115600-28d1-11eb-8e5e-f9d341ee3f64.png)
Here's what it logs in the console after a failed login (serveur returning 401):
![image](https://user-images.githubusercontent.com/1426357/99385899-e9c1cc00-28d1-11eb-8e53-cdaa5da5c3f4.png)

For the record, this is my auth config:
```javascript
Vue.axios = storeApi;
Vue.router = router;
Vue.use(VueAuth, {
  auth,
  http,
  router: authRouter,
  authRedirect: false,
  tokenDefaultKey: 'tzauth',
  loginData: {
    url: '/login',
    method: 'POST',
    redirect: false,
    fetchUser: false,
    staySignedIn: true,
  },
  logoutData: {
    redirect: false,
  },
  registerData: {
    url: '/register',
    method: 'POST',
    redirect: false,
    autoLogin: false,
    staySignedIn: true,
  },
  fetchData: {
    url: '/customers/me',
    method: 'GET',
    enabled: false,
  },
  refreshData: {
    enabled: false,
  },
});
```

This is the problem in your code:
https://github.com/websanova/vue-auth/blob/c06925107499d54e0c2b906e454709d370673359/src/auth.js#L196
This line is executed and creates an unecessary redirect because the login failed, it was called by this:
https://github.com/websanova/vue-auth/blob/c06925107499d54e0c2b906e454709d370673359/src/auth.js#L180
Which is the interceptor of the response.
The issue here is, at no point are you cheking the config at any point and again, you are doing a push with a redirect object and despite that your took `route.query` in account here: https://github.com/websanova/vue-auth/blob/c06925107499d54e0c2b906e454709d370673359/src/auth.js#L204
As vue-router's maintainer says, it's not supposed to be used like that.

I don't really know how I can be more clear here.

**Comment by darkylmnx** [comment]
> As for dealing with the 401, I think you should be able to control that via the http driver.
> 
> https://github.com/websanova/vue-auth/blob/master/src/drivers/http/axios.1.x.js
> https://github.com/websanova/vue-auth/blob/master/src/drivers/http/vue-resource.1.x.js
> 
> There is an `invalidToken` method you can override.

The 401 in the first place shouldn't trigger a redirect when all redirect options where set to false.
The 401 isn't even the issue here, the issue is just that your call a `_processInvalidToken` in which you call a `_processLogout` without checking the `logoutData` config which is supposed to be for that.

**Comment by websanova** [comment]
I think you are misunderstanding what that redirect option is for. It's meant for something like a redirect after a login success, etc. Its just a quick way to do a roiter.push.

And as mentioned if you don't like the redirect occurring on a 401 response you can just override that driver method.

**Comment by darkylmnx** [comment]
I'm not misunderstanding, I'm just telling you what one would assume after knowing that you programmatically logout when there's a 401.

> And as mentioned if you don't like the redirect occurring on a 401 response you can just override that driver method.

This doesn't solve anything, at this point I could just recreate my own plugin if I had to handle the invalidation myself.

The issue isn't the fact of redirecting, the issue is that on redirects to path with query params like `/something?query=param`, that doesn't work. Why? Because of what I explained twice earlier and what Posva answered on the Vue-router issue.
